### PR TITLE
refactor: move texture component to own assembly

### PR DESCRIPTION
**Pull request by BrianAmadori**
## What does this PR change?

DCLTexture is tightly coupled to other unrelated components, like UIShape, VideoTexture, PBRMaterial, BasicMaterial, and so on.

This will not scale well, because:
* Components should be implemented as Plugins on the future, and for this purpose, the Texture component can't be part of MainScripts assembly.
* All the SDK UI system will be reimplemented, and for this initiative, all the UI components should me moved to their own assemblies to lower the maintenance cost of the overhaul. This means that the UI components should be decoupled from the DCLTexture component.

Also, this is an actionable for the following initiative: https://github.com/decentraland/unity-renderer/issues/1357

--

`AttachTo` and `DetachFrom` now use `ISharedComponent`.

Some pure methods have been moved to the new `ComponentUtils` static class.

## How to test the changes?

1. Go to: https://play.decentraland.zone/?renderer-branch={branch_name}&{desired_url_params}
2. ...

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md

### CODE REVIEW

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Textures/ComponentTextureAttachment.cs [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+                        entity.OnShapeUpdated -= OnEntityShapeUpdated;
+                    }
+                }
+            }
+        }
+
+        private void SubscribeToEntityUpdates()
+        {
+            if (disposableComponent.attachedEntities.Count > 0)
+            {
```
Comment (AjimenezDCL) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Textures/ComponentTextureAttachment.cs:188]: Can we invert this logic for an early return and avoid nesting?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Textures/ComponentTextureAttachment.cs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        private Action<ITextureAttachment> OnUpdateEvent;
+
+        // As the SubscribeToEntityUpdates() method is not easy on the CPU, we only call it if the OnUpdateEvent
+        // is actually used. this means that DCLVideoTexture will still work without incurring overhead to
+        // DCLTexture components.
+        public event Action<ITextureAttachment> OnUpdate
+        {
+            add
+            {
+                if ( OnUpdateEvent == null || OnUpdateEvent.GetInvocationList().Length == 0 )
```
Comment (AjimenezDCL) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Textures/ComponentTextureAttachment.cs:22]: What happens if the disposableComponent gets attached to a new entity but `OnUpdate +=` is not called again? We will miss an update from that entity?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Materials/BasicMaterial.cs [hunk 3]
```diff
@@ -83,12 +83,12 @@ public override IEnumerator ApplyChanges(BaseModel newModel)
                         {
                             if ( dclTexture != null )
                             {
-                                dclTexture.DetachFrom(this);
+                                dclTexture.DetachFrom(new ComponentTextureAttachment(this));
```
Comment (AjimenezDCL) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Materials/BasicMaterial.cs:86]: What are the upsides of using a new `ComponentTextureAttachment` instance every time?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Textures/DCLTexture.cs [hunk 4]
```diff
@@ -149,53 +151,55 @@ public override IEnumerator ApplyChanges(BaseModel newModel)
...
+        }
 
-        public virtual void DetachFrom(BasicMaterial component) => RemoveReference(component);
+        public virtual void DetachFrom(ITextureAttachment attachment = null)
+        {
+            RemoveReference(attachment);
+        }
 
-        public virtual void DetachFrom(UIImage component) => RemoveReference(component);
+        protected int refCount;
```
[the remark this answers is not in the document]
Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Textures/DCLTexture.cs:164]: If needed, shouldnt it be located after the `ContainsKey` check?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIImage/UIImage.cs [hunk 5]
```diff
@@ -122,7 +125,7 @@ public override void Dispose()
                 fetchRoutine = null;
             }
 
-            dclTexture?.DetachFrom(this);
+            dclTexture?.DetachFrom(new ComponentTextureAttachment(this));
```
Comment (AjimenezDCL) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIImage/UIImage.cs:128]: Shouldn't this be a `UIShapeTextureAttachment`?

**Review by AjimenezDCL — APPROVED** [review]
I left a few questions, nothing blocking.

To be honest I find the solution a bit too complex, but I've no alternative. Maybe we could add a comment in DCLTexture or the attachments explicitly explaining the approach?

**Review by pravusjif — COMMENTED** [review]
The build appears to work fine with textures of all kinds, although I share the same concerns regarding the constant usage of `new ComponentTextureAttachment()`

File: unity-renderer/Assets/Scripts/MainScripts/DCL/WorldRuntime/Utils/ComponentUtils.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
+﻿using DCL.Components;
+using DCL.Models;
+using UnityEngine;
+
+namespace DCL.Controllers
+{
+    public static class ComponentUtils
```
Comment (AjimenezDCL) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/WorldRuntime/Utils/ComponentUtils.cs:7]: Can we add some early returns on all the methods, I find the nesting more difficult to read

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Textures/DCLTexture.cs [hunk 7]
```diff
@@ -149,38 +151,45 @@ public override IEnumerator ApplyChanges(BaseModel newModel)
...
+            RemoveReference(component);
+        }
 
-        public virtual void DetachFrom(UIImage component) => RemoveReference(component);
+        protected int refCount;
 
-        void AddReference(ISharedComponent component)
+        public void AddReference(ISharedComponent component)
         {
+            refCount++;
```
Comment (AjimenezDCL) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Textures/DCLTexture.cs:168]: Do we need this?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/WorldRuntime/Utils/ComponentUtils.cs (hunk 6 above)
Comment (BrianAmadori) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/WorldRuntime/Utils/ComponentUtils.cs:7]: Done

**Pull request merged**